### ISSUE TITLE
'SAME' padding works incorrect

### ISSUE THREAD

**Issue by VaWheel** [tensorflow/tensorflow#4297]
I have 28x28 image. When I apply convolution (or pooling) with 'SAME' padding, kernel size 2x2, stride 2, it produces feature map of size 14x14, but it should be 15x15, so the last (bottom and right) image pixels would't be processed. The output size should be calculated with formula output = [(input+2*padding-kernel) / stride] + 1 (look https://arxiv.org/pdf/1603.07285.pdf). In case of 'SAME', padding size would be [kernel/2]
TF version 0.9.0

**Comment by tatatodd** [comment]
@VaWheel can you provide a small test example that reproduces your problem?  Thanks!

**Comment by VaWheel** [comment]
`import tensorflow as tf

data = tf.placeholder("float", shape=[1, 8, 8, 1])
conv = tf.nn.max_pool(data, [1, 2, 2, 1], [1, 2, 2, 1], padding='SAME')

print(conv.get_shape())  # prints (1, 4, 4, 1), but should be (1, 5, 5, 1)
`

**Comment by tatatodd** [comment]
@VaWheel I think I see where the confusion occurred.

In the paper, I believe the output formula you're using is is from section 2.4 "Zero padding, non-unit strides".  That formula is correct, but make sure you don't miss the floor function!

But you mention that SAME padding should be [kernel/2], which you're getting from section 2.2.1 "Half (same) padding".  Note that in the formula s=1, i.e. there is a unit stride.  All of section 2.2 is only discussing the s=1 case, so it's not applicable to your stride=2 scenario.  It doesn't make sense to combine the two formulas.

Also note that the use of "same" in section 2.2.1 is different than the terminology we use in tensorflow.  Here's how we define SAME and VALID padding in tensorflow:
https://www.tensorflow.org/versions/r0.10/api_docs/python/nn.html#convolution

Roughly speaking VALID means we never pad, and only apply the filter (i.e. kernel) over real inputs.  SAME means we pad equally on all sides, such that the output size = ceil(input/stride)

Let's consider your original example of a 28x28 image, kernel size 2x2, stride 2.  Starting from the top left corner (0,0), we will move to (0,2), (0,4), ..., (0,26).  That's a total of 14 steps, which is why the output is 14x14.  There are no missed pixels.

Hope that helps.

**Comment by VaWheel** [comment]
As I understand, SAME pads with half the filter size. So in the case of 8x8 image, 2x2 kernel, stride 2 it should look like
![image](https://cloud.githubusercontent.com/assets/12466121/18409407/fb3d885e-774e-11e6-873f-4cfff385b62c.png)

**Comment by VaWheel** [comment]
@tatatodd, what do you think?

**Comment by tatatodd** [comment]
@VaWheel please re-read my explanation.  You're deriving the definition of "SAME" from the paper you referenced, section 2.2.1 "Half (same) padding", which is only valid for unit strides, i.e. s=1.  It is not valid for any other stride.  It is also not the same meaning that we use in tensorflow for the SAME padding.

In tensorflow for an 8x8 image, 2x2 kernel, stride 2 with SAME padding, we will end up with no padding.

**Comment by VaWheel** [comment]
Ok, no padding - that's what I've asked. Thanks

**Comment by zhimengfan1990** [comment]
I got the similar issue. In my case, input shape is (3, 75, 75), the output shape is (3, 37, 37). But according to the formula: https://www.tensorflow.org/api_docs/python/tf/nn/convolution, the output shape should be (3, 38, 38).

**Comment by KHBillel** [comment]
Zero padding layer can help you with this issue.
